### DISCUSSION TITLE
Compiling code for genericSTM32F411CE via platformio fails with an error for stm32duino (2.8.1)

### DISCUSSION THREAD

**Discussion by brightproject** [stm32duino/Arduino_Core_STM32#2523] [Q&A] [answered]
I am compiling code for `blackpill_f411ce` microcontroller.
I get a compiler error:

> Processing genericSTM32F411CE (platform: ststm32; board: blackpill_f411ce; framework: arduino)
> Verbose mode can be enabled via `-v, --verbose` option
> CONFIGURATION: https://docs.platformio.org/page/boards/ststm32/blackpill_f411ce.html
> PLATFORM: ST STM32 (17.4.0) > WeAct Studio BlackPill V2.0 (STM32F411CE)    
> HARDWARE: STM32F411CEU6 100MHz, 128KB RAM, 512KB Flash
> DEBUG: Current (blackmagic) External (blackmagic, cmsis-dap, jlink, stlink)
> PACKAGES:
>  - framework-arduinoststm32 @ 4.20801.240815 (2.8.1)
>  - framework-cmsis @ 2.50900.0 (5.9.0)
>  - toolchain-gccarmnoneeabi @ 1.120301.0 (12.3.1)
> LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
> LDF Modes: Finder ~ chain, Compatibility ~ soft
> Found 19 compatible libraries
> Scanning dependencies...
> Dependency Graph
> |-- CalLib
> |-- I2CDev
> |-- RTIMULib
> |-- RTArduLink
> |-- SimpleKalman
> |-- EEPROM @ 2.0.1
> |-- Wire @ 1.0.0
> Building in release mode
> Linking .pio\build\genericSTM32F411CE\firmware.elf

> .platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/12.3.1/../../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard\libc_nano.a(libc_a-openr.o): in function `_open_r':

> openr.c:(.text._open_r+0x10): undefined reference to `_open'
> collect2.exe: error: ld returned 1 exit status
> *** [.pio\build\genericSTM32F411CE\firmware.elf] Error 1
> ========================= [FAILED] Took 6.20 seconds=========================
> 
> Environment         Status    Duration
> ------------------  --------  ------------
> genericSTM32F411CE  FAILED    00:00:06.195
> ========================= 1 failed, 0 succeeded in 00:00:06.195=========================
>  *  The terminal process ".platformio\penv\Scripts\platformio.exe 'run'" terminated with exit code: 1. 
>  *  Terminal will be reused by tasks, press any key to close it. 

I didn't find any useful advice, there are some explanations, but what needs to be changed and in which file is beyond my knowledge.

> https://stackoverflow.com/questions/65453293/what-are-nosys-nano-rdimon-terms-when-using-arm-gcc

I would be grateful to the community for help with this issue.

P.S. There was another error, but I fixed it using the previous method.

> https://github.com/stm32duino/Arduino_Core_STM32/issues/2475#issue-2459604161

**Comment by fpistm** [comment]
Hi @brightproject 
Unfortunately, we do not support PIO only Arduino IDE 2.x.
You should submit your issue on pio forum or GitHub.

**Reply by brightproject** [reply]
> we do not support PIO only Arduino IDE 2.x.

Sorry @fpistm , but I'm confused.
In `platformio` I use this framework:

> https://registry.platformio.org/tools/platformio/framework-arduinoststm32/installation

And in it the link leads to this repository:

> https://github.com/stm32duino/Arduino_Core_STM32

That's actually why I wrote a question in the discussion.
I'll also try to use the repository from here:

> https://registry.platformio.org/platforms/platformio/ststm32/installation
> https://github.com/platformio/platform-ststm32

Sorry to bother you🙂

**Reply by fpistm** [reply]
Yes they integrate this core but the integration is different than then one with Arduino IDE.  If you reproduced using Arduino IDE the you can submit here. But I can't support the integration from pio.

**Reply by brightproject** [reply]
> If you reproduced using Arduino IDE then you can submit here.

Yes, now I understand the difference.
Thanks again and sorry for the inconvenience.

**Reply by fpistm** [reply]
Welcome 😉

**Reply by brightproject** [reply]
> Welcome

Thanks @fpistm 

The error is similar in the `Arduino IDE` environment.

![compile_error](https://github.com/user-attachments/assets/cc1161e3-5be3-4231-8b32-2dd943ef16d1)

> arm-none-eabi/lib/thumb/v7e-m+fp/hard\libc_nano.a(libc_a-openr.o): in function `_open_r':
> openr.c:(.text._open_r+0x10): undefined reference to `_open'
> collect2.exe: error: ld returned 1 exit status

The source code is in the archive.

[RTIMULib2_New.zip](https://github.com/user-attachments/files/17137421/RTIMULib2_New.zip)

If only I knew the reason for such an error during assembly, I would refactor the code.

**Comment by brightproject** [comment, marked as answer]
Found the reason for the compilation errors, very strange and incomprehensible to me.
Answer received by indirect signs, thanks to information from here:

> https://devzone.nordicsemi.com/f/nordic-q-a/105594/undefined-reference-to-_open-when-using-libc-with-ncs-2-4-2-and-above

In [actual](https://github.com/orgs/stm32duino/discussions/2523) discussion it was pointed out that the problem may be the use of `fopen()`, in the specific code

> https://github.com/avpripri/RTIMULib2

`fopen` is used to open the `*ini` file with settings, for the case of using the code on the `RPI` platform with `Linux`.

Just commented out the lines.

> https://github.com/avpripri/RTIMULib2/blob/3ff3a59101eb09523a2e375b5c43edb98fce77da/RTIMULib/RTIMUSettings.cpp#L604
> 
> https://github.com/avpripri/RTIMULib2/blob/3ff3a59101eb09523a2e375b5c43edb98fce77da/RTIMULib/RTIMUSettings.cpp#L910

![compile_good](https://github.com/user-attachments/assets/882b47db-2f11-4172-93d1-a408384daac1)

The code compiles quickly and without errors.
Also builds for `.pio`

> https://github.com/platformio/platform-ststm32/issues/801#issuecomment-2377561580

![STM32F411_AHRS_IMU_COMPILE_OK](https://github.com/user-attachments/assets/888ab16c-e24f-41d8-966a-8033fec2de8e)

The error is extremely unclear and perhaps it will help someone in the future.
